src/Graphex/Cabal.hs

```diff
@@ -172,21 +183,44 @@ discoversUnit = curry $ \case
 data CabalDiscoverOpts = CabalDiscoverOpts
   { toDiscover      :: NonEmpty CabalDiscoverType
   , includeExternal :: Bool
-  } deriving stock (Show, Eq)
+  , numJobs         :: Int
+  , pruneTo         :: Maybe (ModuleName -> Bool)
+  }
 
 discoverCabalModuleGraph :: CabalDiscoverOpts -> IO ModuleGraph
 discoverCabalModuleGraph opts@CabalDiscoverOpts{..} = do
-  fs <- getDirectoryContents "." -- XXX
-  mods <- fmap fold . pooledMapConcurrently (discoverCabalModules opts) . filter ((".cabal" ==) . takeExtension) $ fs
-
-  let modSet = foldMap (Set.singleton . name) mods
-  gs <- pooledForConcurrently mods $ \Module{..} -> case path of
-    ModuleFile modPath -> do
-      allImps <- parseFileImports modPath
-      let filteredImps =
-            if includeExternal
-            then allImps
-            else filter (\Import{..} -> Set.member module_ modSet) allImps
-      pure $ mkModuleGraph name $ fmap module_ filteredImps
-    ModuleNoFile -> pure $ mkModuleGraph name mempty
-  pure $ fold gs
+  fs <- getDirectoryContents "."
+  mods <- fmap fold . traverse (discoverCabalModules opts) . filter ((".cabal" ==) . takeExtension) $ fs
+
+  let modMap = foldMap (\m@Module{..} -> Map.singleton name m) mods
+  case pruneTo of
+    Nothing -> do
+      gs <- pooledForConcurrentlyN numJobs mods $ \Module{..} -> case path of
+        ModuleFile modPath -> do
+          allImps <- parseFileImports modPath
+          let filteredImps = filter (\Import{..} -> Map.member module_ modMap) allImps
+          pure $ mkModuleGraph name $ fmap module_ (if includeExternal then allImps else filteredImps)
+        ModuleNoFile -> pure $ mkModuleGraph name mempty
+      pure $ fold gs
+    Just keep -> do
+      sem <- newQSem numJobs
+      graphRef <- newTVarIO @ModuleGraph mempty
+      seenRef <- newTVarIO @(Set.Set ModuleName) mempty
+      let go Module{..} = case path of
+            ModuleFile modPath -> do
+              importedMods <- bracket_ (waitQSem sem) (signalQSem sem) $ do
+                seen <- (Set.member name) <$> readTVarIO seenRef
+                if seen then pure [] else do
+                  logit $ unwords ["Start parsing imports for", T.unpack $ unModuleName name]
+                  allImps <- parseFileImports modPath
+                  logit $ unwords ["Finished parsing imports for", T.unpack $ unModuleName name]
+                  let filteredImps = filter (\Import{..} -> Map.member module_ modMap) allImps
+                  atomically $ modifyTVar' graphRef (\g -> mappend g $ mkModuleGraph name $ fmap module_ (if includeExternal then allImps else filteredImps))
+                  atomically $ modifyTVar' seenRef (Set.insert name)
+                  pure $ mapMaybe (\Import{..} -> Map.lookup module_ modMap) allImps
+              mapConcurrently_ go importedMods
```

Comment:
We do this _outside_ the scope of the semaphore. So we

1. Do our work with the semaphore
2. Release the semaphore
3. Handle our children